### PR TITLE
[BEAM-8651] [BEAM-8874] Change pickle_lock to be a reentrant lock, and use it on Py3 only.

### PR DESCRIPTION
**Pull request by tvalentyn**
Unpickling can result in module imports, and module imports can instantiate Beam objects in global scope, such as beam.DoFn(), which call `pickler.loads/dumps` methods, and cause a deadlock in pickle_lock.

Changing the lock to be a re-entrant lock, but removing locks on Python 2 since RLock on Python 2 is slow (https://bugs.python.org/issue3001), and we have not seen a need to have a lock on Python 2 in this codepath.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by tvalentyn** [comment]
R: @aaltay 
cc: @kennknowles @robertwb @Ardagan

**Comment by tvalentyn** [comment]
Run Dataflow Python ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Window tests in VR suite, now passing, however another test failed: apache_beam.transforms.ptransform_test.PTransformTest.test_do_fn_with_windowing_in_finish_bundle , likely a preexisting failure.

File: sdks/python/apache_beam/internal/pickler.py [hunk 1]
```diff
@@ -40,9 +40,22 @@
...
+
+class _NoOpContextManager(object):
+  def __enter__(self):
+    pass
+
+  def __exit__(self, *unused_exc_info):
+    pass
+
+
+if sys.version_info[0] > 2:
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/pickler.py]: Add a TODO to remove this after py2 support is removed?

File: sdks/python/apache_beam/internal/pickler.py [hunk 2]
```diff
@@ -40,9 +40,22 @@
...
+  # Pickling, especially unpickling, causes broken module imports on Python 3
+  # if executed concurrently, see: BEAM-8651, http://bugs.python.org/issue38884.
+  pickle_lock_unless_py2 = threading.RLock()
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/pickler.py]: Have you mentioned perf issues related to RLock? If yes, should we add a link to that?

Comment (tvalentyn) [reply sdks/python/apache_beam/internal/pickler.py]: It's supposedly only below Python 3.3: https://bugs.python.org/issue3001 (currently the link is in PR).

**Comment by tvalentyn** [comment]
Failing VR tests have been reverted in https://github.com/apache/beam/pull/10262.

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Pull request merged**